server/jetty/src/main/java/io/deephaven/server/jetty/JsPlugins.java

```diff
@@ -1,69 +1,107 @@
-import org.eclipse.jetty.util.resource.PathResource;
-import org.eclipse.jetty.util.resource.Resource;
-import org.eclipse.jetty.webapp.WebAppContext;
+import io.deephaven.plugin.js.JsPlugin;
+import io.deephaven.plugin.js.JsPlugin.Visitor;
+import io.deephaven.plugin.js.JsPluginManifestPath;
+import io.deephaven.plugin.js.JsPluginPackagePath;
 
 import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.net.URI;
 import java.nio.file.Files;
-import java.nio.file.Path;
-import java.util.Optional;
+import java.util.Objects;
 import java.util.function.Consumer;
 
-import static org.eclipse.jetty.servlet.ServletContextHandler.NO_SESSIONS;
+import static io.deephaven.server.jetty.Json.OBJECT_MAPPER;
 
-class JsPlugins {
+class JsPlugins implements Consumer<JsPlugin> {
+    static final String JS_PLUGINS = "js-plugins";
 
-    private static final Logger log = LoggerFactory.getLogger(JsPlugins.class);
+    static JsPlugins create() throws IOException {
+        return new JsPlugins(JsPluginsZipFilesystem.create());
+    }
+
+    private final JsPluginsZipFilesystem zipFs;
 
-    public static void maybeAdd(Consumer<Handler> addHandler) {
-        resource()
-                .map(ControlledCacheResource::wrap)
-                .ifPresent(resource -> addResource(addHandler, resource));
+    private JsPlugins(JsPluginsZipFilesystem zipFs) {
+        this.zipFs = Objects.requireNonNull(zipFs);
     }
 
-    private static void addResource(Consumer<Handler> addHandler, ControlledCacheResource resource) {
-        log.info().append("Creating JsPlugins context with resource '").append(resource.toString()).append('\'').endl();
-        WebAppContext context =
-                new WebAppContext(null, "/js-plugins/", null, null, null, new ErrorPageErrorHandler(), NO_SESSIONS);
-        context.setBaseResource(resource);
-        context.setInitParameter(DefaultServlet.CONTEXT_INIT + "dirAllowed", "false");
-        // Suppress warnings about security handlers
-        context.setSecurityHandler(new ConstraintSecurityHandler());
-        addHandler.accept(context);
+    public URI filesystem() {
+        return zipFs.filesystem();
     }
 
-    private static Optional<Resource> resource() {
-        // Note: this would probably be better to live in JettyConfig - but until we establish more formal expectations
-        // for js plugin configuration and workflows, we'll keep this here.
-        final String resourceBase =
-                Configuration.getInstance().getStringWithDefault("deephaven.jsPlugins.resourceBase", null);
-        if (resourceBase == null) {
-            // defaults to "<configDir>/js-plugins/" if it exists
-            return defaultJsPluginsDirectory()
-                    .filter(Files::exists)
-                    .map(PathResource::new);
+    @Override
+    public void accept(JsPlugin jsPlugin) {
+        final IOException ioException = jsPlugin.walk(new CopyToZipFilesystem());
```

Comment:
It's a fair criticism. I wish there was an easy way to scope things as "anything in this package can use this, but nothing else". Will update.

Reply:
Right now, `JsPluginsZipFilesystem` is in the jetty server package, so we can't easily attach the copy logic into the `io.deephaven.plugin.js` classes. I'm hoping the de-Visitor-ifications is satisfactory enough in simplifying these a bit.

Reply:
The purpose of JMPS is "anything in this ~jar can use this, but nothing else", which is pretty close? That said, we don't use it (..and we're not alone there) since it requires basically everything else to use it too.